File: apps/ios/Janark/UI/Auth/AuthController.swift
import Foundation
import Combine

struct AuthPrompt {
    let reason: String
    let onSuccess: () -> Void
}

@MainActor
final class AuthController: ObservableObject {
    @Published private(set) var prompt: AuthPrompt?

    func require(_ reason: String, onSuccess: @escaping () -> Void = {}) {
        prompt = AuthPrompt(reason: reason, onSuccess: onSuccess)
    }

    func dismiss() {
        prompt = nil
    }

    func consumeSuccess() {
        let p = prompt
        prompt = nil
        p?.onSuccess()
    }
}

File: apps/android/app/src/main/java/org/janark/app/ui/auth/AuthController.kt
package org.janark.app.ui.auth

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

data class AuthPrompt(
    val reason: String,
    val onSuccess: () -> Unit,
)

class AuthController {
    private val _prompt = MutableStateFlow<AuthPrompt?>(null)
    val prompt: StateFlow<AuthPrompt?> = _prompt.asStateFlow()

    fun require(reason: String, onSuccess: () -> Unit = {}) {
        _prompt.value = AuthPrompt(reason, onSuccess)
    }

    fun dismiss() {
        _prompt.value = null
    }

    fun consumeSuccess() {
        val p = _prompt.value
        _prompt.value = null
        p?.onSuccess?.invoke()
    }
}
